# Dog Rates Project
 The dataset that I will be wrangling, analyzing, and visualizing is the tweet archive of Twitter user [@dog_rates](https://twitter.com/dog_rates), also known as [WeRateDogs](https://en.wikipedia.org/wiki/WeRateDogs). WeRateDogs is a Twitter account that rates people's dogs with a humorous comment about the dog. These ratings almost always have a denominator of 10. The numerators, though? Almost always greater than 10. 11/10, 12/10, 13/10, etc. Why? Because "[they're good dogs Brent.](https://knowyourmeme.com/memes/theyre-good-dogs-brent)" WeRateDogs has over 4 million followers and has received international media coverage.

WeRateDogs downloaded their Twitter archive and sent it to Udacity via email exclusively for I to use in this project. This archive contains basic tweet data (tweet ID, timestamp, text, etc.) for all 5000+ of their tweets as they stood on August 1, 2017. More on this soon.

Image Predictions File
One more cool thing: Udacity ran every image in the WeRateDogs Twitter archive through a neural network that can classify breeds of dogs*. The results: a table full of image predictions (the top three only) alongside each tweet ID, image URL, and the image number that corresponded to the most confident prediction (numbered 1 to 4 since tweets can have up to four images) and all of them on this URL "https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv".

Each tweet's retweet count and favorite ("like") count at minimum, and any additional data I find interesting. Using the tweet IDs in the WeRateDogs Twitter archive, query the Twitter API for each tweet's JSON data using Python's Tweepy library and store each tweet's entire set of JSON data in a file called 'tweet_json.txt' file. 

## Gather


In [1]:
from io import StringIO
import pandas as pd
import numpy as np
import requests 
import tweepy
import json

In [2]:
#First I load the Twitter archive csv as data frame twitter_archive
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
#Second I load Image predictionis file from the url using requests library 
response = requests.get("https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv")
res = response.content
image_predictions = pd.read_csv(StringIO(res.decode()), sep= '\t')

In [ ]:
# Third data I load each tweet's retweet cont and favorite count at minimu and any additional data I finding interesting using Twitter API "Tweepy"
from tweepy import OAuthHandler
from timeit import default_timer as timer

# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
# These are hidden to comply with Twitter's API terms and conditions

consumer_key = 'HIDDEN'
consumer_secret = 'HIDDEN'
access_token = 'HIDDEN'
access_secret = 'HIDDEN'

auth = OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True)

# NOTE TO REVIEWER: this student had mobile verification issues so the following
# Twitter API code was sent to this student from a Udacity instructor
# Tweet IDs for which to gather additional data via Twitter's API

tweet_ids = twitter_archive.tweet_id.values
len(tweet_ids)

# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
count = 0
fails_dict = {}
start = timer()
# Save each tweet's returned JSON as a new line in a .txt file
with open('tweet_json.txt', 'w') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
    for tweet_id in tweet_ids:
        count += 1
        print(str(count) + ": " + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            print("Success")
            json.dump(tweet._json, outfile)
            outfile.write('\n')
        except tweepy.TweepError as e:
            print("Fail")
            fails_dict[tweet_id] = e
            pass
end = timer()
print(end - start)
print(fails_dict)

In [6]:
#get tweet_id, retweet_count, and favorite_count from text "tweet_json.txt" and put them in data frame call "tweet_json"
tweet_json = []
with open('tweet_json.txt', 'r') as infile:
    for line in infile:
        line = json.loads(line)
        tweet_id = line['id']
        retweet_count = line['retweet_count']
        favorite_count = line['favorite_count']
        tweet_json.append({'tweet_id': tweet_id, 'retweet_count': retweet_count,'favorite_count': favorite_count})
tweet_json = pd.DataFrame(tweet_json)

## Assess
#### Visual Assessment

In [7]:
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [8]:
twitter_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [9]:
twitter_archive.duplicated().sum()

0

In [10]:
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [11]:
image_predictions.duplicated().sum()

0

In [12]:
image_predictions

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [13]:
tweet_json.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
favorite_count    2354 non-null int64
retweet_count     2354 non-null int64
tweet_id          2354 non-null int64
dtypes: int64(3)
memory usage: 55.2 KB


In [14]:
tweet_json

favorite_count  retweet_count            tweet_id
0              39467           8853  892420643555336193
1              33819           6514  892177421306343426
2              25461           4328  891815181378084864
3              42908           8964  891689557279858688
4              41048           9774  891327558926688256
5              20562           3261  891087950875897856
6              12041           2158  890971913173991426
7              56848          16716  890729181411237888
8              28226           4429  890609185150312448
9              32467           7711  890240255349198849
10             31166           7624  890006608113172480
11             28268           5156  889880896479866881
12             38818           8538  889665388333682689
13             27672           4735  889638837579907072
14             15359           2321  889531135344209921
15             25652           5637  889278841981685760
16             29611           4709  888917238123831296
17             26080           4559  888804989199671297
18             20290           3732  888554962724278272
19             22201           3653  888078434458587136
20             30779           5609  887705289381826560
21             46959          12082  887517139158093824
22             69871          18781  887473957103951883
23             34222          10737  887343217045368832
24             31061           6167  887101392804085760
25             35859           8084  886983233522544640
26             12306           3443  886736880519319552
27             22798           4610  886680336477933568
28             21524           3316  886366144734445568
29               117              4  886267009285017600
...              ...            ...                 ...
2324             459            339  666411507551481857
2325             113             44  666407126856765440
2326             172             92  666396247373291520
2327             194            100  666373753744588802
2328             804            595  666362758909284353
2329             229             77  666353288456101888
2330             307            146  666345417576210432
2331             204             96  666337882303524864
2332             522            368  666293911632134144
2333             152             71  666287406224695296
2334             184             82  666273097616637952
2335             108             37  666268910803644416
2336           14765           6871  666104133288665088
2337              81             16  666102155909144576
2338             164             73  666099513787052032
2339             169             79  666094000022159362
2340             121             47  666082916733198337
2341             335            174  666073100786774016
2342             154             67  666071193221509120
2343             496            232  666063827256086533
2344             115             61  666058600524156928
2345             304            146  666057090499244032
2346             448            261  666055525042405380
2347            1253            879  666051853826850816
2348             136             60  666050758794694657
2349             111             41  666049248165822465
2350             311            147  666044226329800704
2351             128             47  666033412701032449
2352             132             48  666029285002620928
2353            2535            532  666020888022790149

[2354 rows x 3 columns]

#### Programmatic Assessmen

In [15]:
twitter_archive.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2089  670789397210615808                    NaN                  NaN   
1885  674793399141146624           6.717299e+17         4.196984e+09   
840   767122157629476866                    NaN                  NaN   
717   783695101801398276                    NaN                  NaN   
976   750026558547456000                    NaN                  NaN   

                      timestamp  \
2089  2015-11-29 02:20:29 +0000   
1885  2015-12-10 03:30:58 +0000   
840   2016-08-20 22:12:29 +0000   
717   2016-10-05 15:47:27 +0000   
976   2016-07-04 18:00:41 +0000   

                                                 source  \
2089  <a href="http://twitter.com/download/iphone" r...   
1885  <a href="http://twitter.com/download/iphone" r...   
840   <a href="http://twitter.com/download/iphone" r...   
717   <a href="http://twitter.com/download/iphone" r...   
976   <a href="https://about.twitter.com/products/tw...   

                                                   text  retweeted_status_id  \
2089  Two obedient dogs here. Left one has extra leg...                  NaN   
1885  I have found another. 13/10 https://t.co/HwroP...                  NaN   
840   This is Rupert. You betrayed him with bath tim...                  NaN   
717   This is Loomis. He's the leader of the Kenneth...                  NaN   
976   Meet Jax &amp; Jil. Jil is yelling the pledge ...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2089                       NaN                        NaN   
1885                       NaN                        NaN   
840                        NaN                        NaN   
717                        NaN                        NaN   
976                        NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2089  https://twitter.com/dog_rates/status/670789397...                 9   
1885  https://twitter.com/dog_rates/status/674793399...                13   
840   https://twitter.com/dog_rates/status/767122157...                13   
717   https://twitter.com/dog_rates/status/783695101...                12   
976   https://twitter.com/dog_rates/status/750026558...                10   

      rating_denominator    name doggo floofer pupper puppo  
2089                  10    None  None    None   None  None  
1885                  10    None  None    None   None  None  
840                   10  Rupert  None    None   None  None  
717                   10  Loomis  None    None   None  None  
976                   10     Jax  None    None   None  None

In [16]:
twitter_archive.query("doggo != 'None'")

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
9     890240255349198849                    NaN                  NaN   
43    884162670584377345                    NaN                  NaN   
99    872967104147763200                    NaN                  NaN   
108   871515927908634625                    NaN                  NaN   
110   871102520638267392                    NaN                  NaN   
121   869596645499047938                    NaN                  NaN   
172   858843525470990336                    NaN                  NaN   
191   855851453814013952                    NaN                  NaN   
200   854010172552949760                    NaN                  NaN   
211   851953902622658560                    NaN                  NaN   
240   846514051647705089                    NaN                  NaN   
248   845397057150107648                    NaN                  NaN   
300   836753516572119041                    NaN                  NaN   
318   834574053763584002                    NaN                  NaN   
323   834089966724603904                    NaN                  NaN   
331   832998151111966721                    NaN                  NaN   
339   832273440279240704                    NaN                  NaN   
344   832032802820481025                    NaN                  NaN   
345   831939777352105988                    NaN                  NaN   
351   831322785565769729                    NaN                  NaN   
359   829878982036299777                    NaN                  NaN   
362   829449946868879360                    NaN                  NaN   
363   829374341691346946                    NaN                  NaN   
372   828381636999917570                    NaN                  NaN   
384   826958653328592898                    NaN                  NaN   
385   826848821049180160                    NaN                  NaN   
389   826476773533745153                    NaN                  NaN   
391   826204788643753985                    NaN                  NaN   
423   821765923262631936                    NaN                  NaN   
425   821421320206483457                    NaN                  NaN   
...                  ...                    ...                  ...   
857   763956972077010945           7.638652e+17         1.584641e+07   
877   760893934457552897                    NaN                  NaN   
881   760521673607086080                    NaN                  NaN   
889   759793422261743616                    NaN                  NaN   
899   758828659922702336                    NaN                  NaN   
914   757393109802180609                    NaN                  NaN   
919   756526248105566208                    NaN                  NaN   
924   755206590534418437                    NaN                  NaN   
944   752682090207055872                    NaN                  NaN   
945   752660715232722944                    NaN                  NaN   
948   752334515931054080                    NaN                  NaN   
956   751583847268179968                    NaN                  NaN   
967   750381685133418496           7.501805e+17         4.717297e+09   
977   750011400160841729                    NaN                  NaN   
985   749075273010798592                    NaN                  NaN   
989   748932637671223296                    NaN                  NaN   
992   748692773788876800                    NaN                  NaN   
1030  745433870967832576                    NaN                  NaN   
1039  744234799360020481                    NaN                  NaN   
1051  742534281772302336                    NaN                  NaN   
1063  741067306818797568                    NaN                  NaN   
1075  739623569819336705                    NaN                  NaN   
1079  739238157791694849                    NaN                  NaN   
1103  735256018284875776                    NaN         

In [17]:
twitter_archive.rating_numerator.sort_values()

315        0
1016       0
2335       1
2261       1
2338       1
605        1
1446       1
1869       1
2091       1
2038       1
1940       1
1764       2
2310       2
1761       2
1920       2
2246       2
2237       2
2349       2
2326       2
2079       2
2305       3
1249       3
1898       3
1478       3
1692       3
2202       3
765        3
1928       3
1314       3
2183       3
        ... 
866       14
83        14
291       15
285       15
55        17
1663      20
516       24
1712      26
763       27
1433      44
1274      45
1202      50
1351      60
340       75
695       75
1254      80
433       84
1843      88
1228      99
1635     121
1634     143
1779     144
902      165
290      182
1120     204
2074     420
188      420
189      666
313      960
979     1776
Name: rating_numerator, Length: 2356, dtype: int64

In [18]:
twitter_archive.rating_denominator.sort_values()

313       0
2335      2
516       7
1576     10
1575     10
1574     10
1573     10
1572     10
1563     10
1570     10
1569     10
1568     10
1567     10
1566     10
1565     10
1564     10
1577     10
1571     10
0        10
1562     10
1547     10
1548     10
1549     10
1550     10
1551     10
1552     10
1553     10
1554     10
1555     10
1556     10
       ... 
782      10
781      10
775      10
799      10
780      10
779      10
776      10
777      10
778      10
800      10
784      11
1068     11
1662     11
342      15
1663     16
1165     20
1598     20
1433     40
1202     50
1351     50
1274     50
433      70
1843     80
1254     80
1228     90
1635    110
1779    120
1634    130
902     150
1120    170
Name: rating_denominator, Length: 2356, dtype: int64

In [19]:
twitter_archive.query('rating_denominator != 10')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
433   820690176645140481                    NaN                  NaN   
516   810984652412424192                    NaN                  NaN   
784   775096608509886464                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
1068  740373189193256964                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1165  722974582966214656                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1635  684222868335505415                    NaN                  NaN   
1662  682962037429899265                    NaN                  NaN   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   

                      timestamp  \
313   2017-02-24 21:54:03 +0000   
342   2017-02-16 04:45:50 +0000   
433   2017-01-15 17:52:40 +0000   
516   2016-12-19 23:06:23 +0000   
784   2016-09-11 22:20:06 +0000   
902   2016-07-28 01:00:57 +0000   
1068  2016-06-08 02:41:38 +0000   
1120  2016-05-13 16:15:54 +0000   
1165  2016-04-21 02:25:47 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1598  2016-01-10 04:04:10 +0000   
1634  2016-01-05 04:11:44 +0000   
1635  2016-01-05 04:00:18 +0000   
1662  2016-01-01 16:30:13 +0000   
1663  2016-01-01 06:22:03 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   
2335  2015-11-16 16:11:11 +0000   

                                                 source  \
313   <a href="http://twitter.com/download/iphone" r...   
342   <a href="http://twitter.com/download/iphone" r...   
433   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
784   <a href="http://twitter.com/download/iphone" r...   
902   <a href="http://twitter.com/download/iphone" r...   
1068  <a href="http://twitter.com/download/iphone" r...   
1120  <a href="http://twitter.com/download/iphone" r...   
1165  <a href="http://twitter.com/download/iphone" r...   
1202  <a href="http://twitter.com/download/iphone" r...   
1228  <a href="http://twitter.com/download/iphone" r...   
1254  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
1433  <a href="http://twitter.com/download/iphone" r...   
1598  <a href="http://twitter.com/download/iphone" r...   
1634  <a href="http://twitter.com/download/iphone" r...   
1635  <a href="http://twitter.com/download/iphone" r...   
1662  <a href="http://twitter.com/download/iphone" r...   
1663  <a href="http://twitter.com/download/iphone" r...   
1779  <a href="http://twitter.com/download/iphone" r...   
1843  <a href="http://twitter.com/download/iphone" r...   
2335  <a href="http://twitter.com/download/iphone" r...   

                             

In [20]:
sum(twitter_archive.expanded_urls.isnull())

59

In [21]:
image_predictions.sample(5)

tweet_id                                            jpg_url  \
162   668955713004314625    https://pbs.twimg.com/media/CUibq3uVAAAup_O.jpg   
464   675003128568291329    https://pbs.twimg.com/media/CV4XwYiWoAAHQIF.jpg   
1864  842892208864923648  https://pbs.twimg.com/ext_tw_video_thumb/80710...   
1968  868552278524837888    https://pbs.twimg.com/media/DA23sCeVoAE3uF0.jpg   
1402  768970937022709760  https://pbs.twimg.com/ext_tw_video_thumb/76896...   

      img_num              p1   p1_conf  p1_dog                 p2   p2_conf  \
162         1  cocker_spaniel  0.367492    True   Lakeland_terrier  0.272621   
464         1        Pembroke  0.655279    True         Pomeranian  0.104164   
1864        1       Chihuahua  0.505370    True         Pomeranian  0.120358   
1968        1         whippet  0.378151    True  Italian_greyhound  0.275935   
1402        1      Pomeranian  0.182358    True   golden_retriever  0.110658   

      p2_dog                              p3   p3_conf  p3_dog  
162     True     soft-coated_wheaten_terrier  0.067006    True  
464     True                        Cardigan  0.052818    True  
1864    True                     toy_terrier  0.077008    True  
1968    True  American_Staffordshire_terrier  0.094991    True  
1402    True                       mousetrap  0.086399   False

In [22]:
image_predictions.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [23]:
tweet_json.sample(5)

favorite_count  retweet_count            tweet_id
884             6581           2382  759943073749200896
2021            9016           4519  671896809300709376
54               129              7  881633300179243008
1389            8282           3129  700143752053182464
168            75193          25661  859196978902773760

In [24]:
tweet_json.describe()

favorite_count  retweet_count      tweet_id
count     2354.000000    2354.000000  2.354000e+03
mean      8080.968564    3164.797366  7.426978e+17
std      11814.771334    5284.770364  6.852812e+16
min          0.000000       0.000000  6.660209e+17
25%       1415.000000     624.500000  6.783975e+17
50%       3603.500000    1473.500000  7.194596e+17
75%      10122.250000    3652.000000  7.993058e+17
max     132810.000000   79515.000000  8.924206e+17

### Quality

#### twitter_archive table
- Check if there missing values in columns(in_reply_to_status_id, in_reply_to_use_id, retweeted_status_id, retweeted_user_id, retweeted_status_timestamp, expanded_urls) from tweet_json.txt
- Fill the missing values in columns(in_reply_to_status_id, in_reply_to_use_id, retweeted_status_id, retweeted_user_id, retweeted_status_timestamp, expanded_urls).
- Erroneous datatypes(tweet_id, in_reply_to_staus_id, in_reply_to_use_id, retweeted_staus_id, retweeted_user_id, retweeted_status_timestamp, timestamp, timestamp)
- Rating numerator have some values are unexpected 20~1776 
- Rating denominator have some values are unexpected equal 2, 0, 7, (11~170)
- different twitter ratios have different ratio between nuerator and denominator.
- Name columns some variables equal "a", "an", "the"

#### image_predictions table
- Erroneous datatypes(tweet_id)

#### tweet_json table
- Missing records(2354 instead of 2356)
- Erroneous datatypes(tweet_id)

### Tidness
- make p1, p2, and p3 in one column in p and p1_conf,p2_conf, and p3_conf in conf column and P1_dog, p2_dog , P3_dog in p_dog column and name the column dog_kind_predictions
- concat tweet_json and twitter_archive table and name it twitter_rate_dog

## Clean

In [226]:
twitter_archive_clean = twitter_archive.copy()
image_predictions_clean = image_predictions.copy()
tweet_json_clean = tweet_json.copy()

### Missing Data
**Check if there missing values in columns(in_reply_to_status_id, in_reply_to_use_id, retweeted_status_id, retweeted_user_id, retweeted_status_timestamp, expanded_urls) from tweet_json.txt**

#### Define

Make dataframe have these values from tweet_json.txt then put the values from the dataframe to twitter_archive dataframe

#### Code

In [227]:
missing_values = []
with open('tweet_json.txt', 'r') as infile:
    for line in infile:
        line = json.loads(line)
        id_ = line['id']
        if line.get('in_reply_to_status_id'):
            in_reply_to_status_id = line['in_reply_to_status_id']
            in_reply_to_user_id = line['in_reply_to_user_id']
        else:
            in_reply_to_status_id = None
            in_reply_to_user_id = None
        
        if line.get('retweeted_status'):
            retweeted_status_id = line['retweeted_status']['id']
            retweeted_user_id =  line['retweeted_status']['user']['id']
            retweeted_status_timestamp = line['retweeted_status']['created_at']
        else:
            retweeted_status_id = None
            retweeted_user_id = None
            retweeted_status_timestamp = None
            
        if line['entities'].get('media'):
            expanded_url = line['entities']['media'][0].get('expanded_url')
        else:
            expanded_url = None
        missing_values.append({'tweet_id': id_,
                               'in_reply_to_status_id': in_reply_to_status_id, 
                               'in_reply_to_user_id': in_reply_to_user_id,
                               'retweeted_status_id': retweeted_status_id,
                               'retweeted_user_id': retweeted_user_id,
                               'retweeted_status_timestamp': retweeted_status_timestamp,
                               'expanded_urls' : expanded_url})
missing_values = pd.DataFrame(missing_values)

#### Test

In [228]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         2356 non-null object
in_reply_to_user_id           2356 non-null object
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           2356 non-null object
retweeted_status_user_id      2356 non-null object
retweeted_status_timestamp    2356 non-null object
expanded_urls                 2356 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: int64(3), object(14)


In [229]:
missing_values.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 7 columns):
expanded_urls                 2073 non-null object
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
retweeted_status_id           179 non-null float64
retweeted_status_timestamp    179 non-null object
retweeted_user_id             179 non-null float64
tweet_id                      2354 non-null int64
dtypes: float64(4), int64(1), object(2)
memory usage: 128.8+ KB


**Fill the missing values in columns(in_reply_to_status_id, in_reply_to_use_id, retweeted_status_id, retweeted_user_id, retweeted_status_timestamp, expanded_urls).**
#### Define
From last error I fill missing values by remove NaN and put 'None' string for all missing values in this columns

#### Code

In [230]:
twitter_archive_clean.fillna('None', inplace= True)

#### Test

In [231]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         2356 non-null object
in_reply_to_user_id           2356 non-null object
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           2356 non-null object
retweeted_status_user_id      2356 non-null object
retweeted_status_timestamp    2356 non-null object
expanded_urls                 2356 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: int64(3), object(14)


### Tidness
**make p1, p2, and p3 in one column in p and p1_conf,p2_conf, and p3_conf in conf column and P1_dog, p2_dog , P3_dog in p_dog column and name the column dog_kind_predictions**

#### Define
make function to calculate for each row the best prediction for the image of dog then apply it for image_predictions_clean and then produce the kind of dog and prediction configuration and drop columns(p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog)

#### Code

In [232]:
def image_pred(row):
    if row['p1_dog'] and row['p1_conf'] >= row['p2_conf'] and row['p1_conf'] >= row['p3_conf']:
        return row['p1'], row['p1_conf'], '1'
    elif row['p2_dog'] and row['p2_conf'] >= row['p3_conf']:
        return row['p2'], row['p2_conf'], '2'
    elif row['p3_dog']:
        return row['p3'], row['p3_conf'], '3'
    else:
        return 'none_dog',None , None
image_predict = image_predictions_clean.apply(image_pred, axis= 1)

In [233]:
image_predictions_clean['prediction_level'] = image_predict.str[2]
image_predictions_clean['dog_kind_prediction'] = image_predict.str[0]
image_predictions_clean['prediction_configuration'] = image_predict.str[1]
image_predictions_clean = image_predictions_clean.drop(columns = [ 'p1', 'p1_conf', 'p1_dog', 'p2',
       'p2_conf', 'p2_dog', 'p3', 'p3_conf', 'p3_dog'])

#### Test

In [234]:
image_predictions_clean

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

**concat tweet_json and twitter_archive table and name it twitter_archive**
#### Define
concat twitter_archive_clean and tweet_json_clean on tweet_id columns one_to_one

#### Code

In [237]:
twitter_archive_clean = pd.merge(twitter_archive_clean, tweet_json_clean, on='tweet_id', how='outer', validate="one_to_one")

#### Test

In [239]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2356 entries, 0 to 2355
Data columns (total 19 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         2356 non-null object
in_reply_to_user_id           2356 non-null object
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           2356 non-null object
retweeted_status_user_id      2356 non-null object
retweeted_status_timestamp    2356 non-null object
expanded_urls                 2356 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
favorite_count               

In [240]:
twitter_archive_clean.head()

tweet_id in_reply_to_status_id in_reply_to_user_id  \
0  892420643555336193                  None                None   
1  892177421306343426                  None                None   
2  891815181378084864                  None                None   
3  891689557279858688                  None                None   
4  891327558926688256                  None                None   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                None   
1  This is Tilly. She's just checking pup on you....                None   
2  This is Archie. He is a rare Norwegian Pouncin...                None   
3  This is Darla. She commenced a snooze mid meal...                None   
4  This is Franklin. He would like you to stop ca...                None   

  retweeted_status_user_id retweeted_status_timestamp  \
0                     None                       None   
1                     None                       None   
2                     None                       None   
3                     None                       None   
4                     None                       None   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  favorite_count  \
0                  10   Phineas  None    None   None  None         39467.0   
1                  10     Tilly  None    None   None  None         33819.0   
2                  10    Archie  None    None   None  None         25461.0   
3                  10     Darla  None    None   None  None         42908.0   
4                  10  Franklin  None    None   None  None         41048.0   

   retweet_count  
0         8853.0  
1         6514.0  
2         4328.0  
3         8964.0  
4         9774.0